internal/sandbox/ocifs/ocifs.go

```diff
@@ -0,0 +1,80 @@
+package ocifs
+
+import (
+	"errors"
+	"io/fs"
+
+	"github.com/stealthrocket/timecraft/internal/sandbox"
+)
+
+const (
+	whiteoutPrefix = ".wh."
+	whiteoutOpaque = ".wh..wh..opq"
+)
+
+// FileSystem is an implementation of the sandbox.FileSystem interface that
+// merges OCI layers into a single view.
+type FileSystem struct {
+	layers []sandbox.FileSystem
+}
+
+// New constructs a file system which combines layers into a flattened view
+// which stacks layers on each other.
+//
+// The returned file system is read-only, it does not allow modifications of
+// the layers that it was constructed from. This includes writes to files as
+// well as modifcations of the directory structure.
+//
+// For the OCI layer specification, see:
+// https://github.com/opencontainers/image-spec/blob/main/layer.md
+func New(layers ...sandbox.FileSystem) *FileSystem {
+	// Reverse the slice of layers so we can use range loops to iterate from
+	// the upper to the lower layers.
+	fsys := &FileSystem{
+		layers: make([]sandbox.FileSystem, len(layers)),
+	}
+	for i, layer := range layers {
+		fsys.layers[len(layers)-(i+1)] = layer
+	}
+	return fsys
+}
+
+// Open satisfies the sandbox.FileSystem interface.
+func (fsys *FileSystem) Open(name string, flags int, mode fs.FileMode) (sandbox.File, error) {
+	f, err := fsys.openRoot()
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return f.Open(name, flags, mode)
+}
+
+func (fsys *FileSystem) openRoot() (*file, error) {
+	if len(fsys.layers) == 0 {
+		return nil, sandbox.ENOENT
+	}
+
+	files := make([]sandbox.File, 0, len(fsys.layers))
+	defer func() {
+		closeFiles(files)
+	}()
```

Comment:
Nit: this pattern of setting `files = nil` right before success always trips me up when reviewing. Could we stick with a convention where cleanup only occurs on error and where you can understand the function line by line without having to read the end first?

```
func (fsys *FileSystem) openRoot() (root *file, err error) {
    // ... 
    
    defer func() {
        if err != nil {
            // cleanup
        }
    }()
    
    // ...
}
```

Reply:
We could but it's a slightly different behavior which assumes that the method either succeeds or errors, it doesn't handle the case where it panics.

The current form will properly close files even if a panic occurs, which can be a more desirable behavior. What do you think?

Reply:
Ok :+1: maybe a comment then in the defer saying that
> `files` is unset if the operation succeeds so that cleanup only occurs on error/panic